tools/asset-sync/assets.ps1

```diff
@@ -0,0 +1,521 @@
+#>
+Function Evaluate-Target-Dir {
+    param (
+        [Parameter(Mandatory = $true)]
+        [string] $TargetPath
+    )
+
+    $isFile = Test-Path $TargetPath -PathType Leaf
+    $foundConfig = $false
+    $foundRoot = $false
+
+    if ($isFile) {
+        throw "Evaluated a file `"$TargetPath`" as a directory. Exiting."
+    }
+
+    # we need to force to ensure we grab hidden files in the directory dump
+    $files = Get-ChildItem -Force $TargetPath
+
+    foreach($file in $files)
+    {
+        if($file.Name.ToLower() -eq "assets.json"){
+            $foundConfig = $true
+        }
+
+        # need to ensure that if we're outside a git repo, we will still break out eventually.
+        if(($file.Name.ToLower() -eq ".git") -or -not (Split-Path $TargetPath)){
+            $foundRoot = $true
+        }
+    }
+
+    return @(
+        $foundConfig, $foundRoot
+    )
+}
+
+<#
+.SYNOPSIS
+Gets the relative path of the assets json within the target repo.
+
+.PARAMETER Config
+The config
+#>
+Function AscendToRepoRoot {
+    param(
+        [Parameter(Mandatory = $false)]
+        [string] $StartPath
+    )
+    $pathForManipulation = $StartPath
+    if(Test-Path $StartPath -PathType Leaf){
+        $pathForManipulation = Split-Path $pathForManipulation
+    }
+
+    $foundConfig, $reachedRoot = Evaluate-Target-Dir -TargetPath $pathForManipulation
+
+    while (-not $reachedRoot){
+        $pathForManipulation, $remainder = Split-Path $pathForManipulation
+
+        $foundConfig, $reachedRoot = Evaluate-Target-Dir -TargetPath $pathForManipulation
+    }
+
+    if ($reachedRoot){
+        return $pathForManipulation
+    }
+    else {
+        throw "Unable to the root of the git repo."
+    }
+}
+
+
+<#
+.SYNOPSIS
+Traverses up from a provided target directory to find a assets JSON, parses it, and returns the location and JSON content.
+
+.DESCRIPTION
+Traverses upwards until it hits either a `.git` folder or a `assets.json` file. Throws an exception if it can't find a assets.json before it hits root.
+
+.PARAMETER TargetPath
+Optional argument specifying the directory to start traversing up from. If not provided, current working directory will be used.
+#>
+Function Resolve-AssetsJson {
+    param (
+        [Parameter(Mandatory = $false)]
+        [string] $TargetPath
+    )
+    $pathForManipulation = $TargetPath
+    $foundConfig = $false
+    $reachedRoot = $false
+
+    if(-not $TargetPath){
+        $pathForManipulation = Get-Location
+    }
+    
+    $foundConfig, $reachedRoot = Evaluate-Target-Dir -TargetPath $pathForManipulation
+
+    while (-not $foundConfig -and -not $reachedRoot){
+        $pathForManipulation, $remainder = Split-Path $pathForManipulation
+
+        $foundConfig, $reachedRoot = Evaluate-Target-Dir -TargetPath $pathForManipulation
+    }
+
+    if ($foundConfig){
+        $discoveredPath = Join-Path $pathForManipulation "assets.json"
+    }
+    else {
+        throw "Unable to locate assets.json"
```

Comment:
This will never hit due to above while loop condition (similar comment as above).

Reply:
I disagree with your assessment on this one.

The while loop here is 

`while(!$foundConfig && !reachedRoot)`

However, we can `reachedRoot` without `foundConfig`. 

In the case where we have reached root, but haven't found config, we have a problem.